modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java

```diff
@@ -55,6 +57,9 @@ public void run(
 		Layout layout = themeDisplay.getLayout();
 
 		if (layout.isTypeControlPanel()) {
+			_setThemeDisplayCSSURLs(
+				httpServletRequest, _getThemeCSSCET(layout), themeDisplay);
```

Comment:
This still doesn't look quite right. Currently I don't think this will cause a bug, but when `layout.isTypeControlPanel() == true` and `Objects.equals(mode, Constants.PREVIEW)` we will call `_setThemeDisplayCSSURLs` twice. First time setting the controlpanel scoped themeCSS and the second time to the themeCSS for the Preview page. This makes the first call unnecessary. Can you find a way to update the logic so we don't call `_setThemeDisplayCSSURLs` twice?

Reply:
@ethib137  I couldn't figure it out a way of not having the call for this twice. But to prevent the execution of this method twice, I have added a condition at line 90

```java
if (!layout.isTypeControlPanel()) {
	_setThemeDisplayCSSURLs(
		httpServletRequest, _getThemeCSSCET(layout), themeDisplay);
}
```

Reply:
Actually, with the condition like this It's going to render the wrong themeCSS on a preview page. It will render the control panel themeCSS but it should render the themeCSS assigned to the page being previewed.

You can fix this by removing the call to `_setThemeDisplayCSSURLs` at line 60 and removing the if statement at line 90 so that we always make the call to `_setThemeDisplayCSSURLs` there.

Reply:
@ethib137 I cannot just delete `_setThemeDisplayCSSURLs` at line 60. Because when the layout is control panel and the p_l_mode is not PREVIEW, it will early return and not apply theme at all

Reply:
@ethib137  Now that you just mentioned, I believe the way to go is to have the call on line 60 and remove the condition on line 90. Because, in case there is a control panel theme and the layout regards the preview of a page, the it will look for the theme applied to that page, so line 90 would override what was set in line 60. But, most of the time, that won't happen, so we do not have performance issues

Reply:
@ethib137 what about

```java
@Override
	public void run(
		HttpServletRequest httpServletRequest,
		HttpServletResponse httpServletResponse) {

		ThemeDisplay themeDisplay =
			(ThemeDisplay)httpServletRequest.getAttribute(
				WebKeys.THEME_DISPLAY);

		Layout layout = themeDisplay.getLayout();

		String mode = ParamUtil.getString(
			httpServletRequest, "p_l_mode", Constants.VIEW);

		if (layout.isTypeControlPanel()) {
			_setThemeDisplayCSSURLs(
				httpServletRequest, _getThemeCSSCET(layout), themeDisplay);

			if (!Objects.equals(mode, Constants.PREVIEW)) {
				return;
			}

			long selPlid = ParamUtil.getLong(
				httpServletRequest,
				StringBundler.concat(
					StringPool.UNDERLINE,
					ParamUtil.getString(httpServletRequest, "p_p_id"),
					"_selPlid"));

			if (selPlid <= 0) {
				return;
			}

			layout = _layoutLocalService.fetchLayout(selPlid);
		}

		if (layout == null) {
			return;
		}

		themeDisplay.setFaviconURL(_getFaviconURL(layout));

		if (!layout.isTypeControlPanel() || Objects.equals(mode, Constants.PREVIEW)) {
			_setThemeDisplayCSSURLs(
				httpServletRequest, _getThemeCSSCET(layout), themeDisplay);
		}

		ThemeSpritemapCET themeSpritemapCET = _getThemeSpritemapCET(layout);

		if (themeSpritemapCET != null) {
			themeDisplay.setPathThemeSpritemap(themeSpritemapCET.getURL());
		}
	}
```